net: add TcpListener.local_addr method

FrankReh opened the pull request:
The local_addr method is a blocking call.

FrankReh commented:
Fixes TCP portion of #106 

But maybe there is a more standard way to cast the fd to a net::TcpListener?

Noah-Kennedy commented on src/net/tcp/listener.rs:
```diff
@@ -40,15 +40,43 @@ impl TcpListener {
 
+    /// Returns the local address that this listener is bound to.
+    ///
+    /// This can be useful, for example, when binding to port 0 to figure out
```
I wouldn't exactly call this blocking. Certain system calls (like this one) are fast enough that they can be considered nonblocking. This one will simply drop to kernelspace to grab some data about the socket and return back.

It's probably even less expensive than another syscall we generally agree to be nonblocking, the `recv` syscall (in nonblocking mode), which does far more work in kernelspace than this one!

FrankReh replied:
Agreed.

Noah-Kennedy commented on src/net/tcp/listener.rs:
```diff
@@ -40,15 +40,43 @@ impl TcpListener {
     ///
     /// Binding with a port number of 0 will request that the OS assigns a port
     /// to this listener.
-    ///
```
Yeah I think this comment was kinda wrong tbh anyways.

Noah-Kennedy commented on src/net/tcp/listener.rs:
````diff
@@ -40,15 +40,43 @@ impl TcpListener {
+    /// let addr = listener.local_addr().expect("Couldn't get local address");
+    /// assert_eq!(addr, SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 8080)));
+    /// ```
+    #[cfg(unix)]
````
Not sure if the `#[cfg(unix)]` is necessary considering that this crate only works on unix in the first place.

FrankReh replied:
I wasn't sure either but thought it safer to include so you could comment. Happy to remove it.

Noah-Kennedy replied:
Yeah I'd recommend removing it

Noah-Kennedy commented:
Some minor feedback, but overall this looks good!

FrankReh commented on src/net/tcp/listener.rs:
```diff
@@ -40,15 +40,43 @@ impl TcpListener {
+        use std::os::unix::io::{AsRawFd, FromRawFd};
+
+        let fd = self.inner.as_raw_fd();
+        // SAFETY: Our fd is the handle the kernel has given us for a TcpListener.
```
Is this SAFETY comment too much? Happy to replace if you want to suggest an alternative.

Noah-Kennedy replied:
This is great!

FrankReh commented:
Think I handled your first set of comments.

I added a question about the SAFETY comment. Is it too wordy?

And importantly, I've added the local_addr method for the UnixListener now but have only tested it with the document example.
